### PR TITLE
Prefix billing

### PR DESCRIPTION
**Pull request by nickpell**
implements http://manage.dimagi.com/default.asp?154612#861986

### CODE REVIEW

File: corehq/apps/smsbillables/models.py [hunk 1]
```diff
@@ -27,23 +27,38 @@ class SmsGatewayFeeCriteria(models.Model):
     backend_instance = models.CharField(max_length=255, db_index=True, null=True)
     direction = models.CharField(max_length=10, db_index=True, choices=DIRECTION_CHOICES)
     country_code = models.IntegerField(max_length=5, null=True, blank=True, db_index=True)
+    prefix = models.CharField(max_length=10, blank=True, default="", db_index=True)
 
     @classmethod
-    def get_most_specific(cls, backend_api_id, direction, backend_instance=None, country_code=None):
+    def get_most_specific(cls, backend_api_id, direction, backend_instance=None, country_code=None, national_number=None):
```
Comment (dimagimon) [inline corehq/apps/smsbillables/models.py]: E501 line too long (122 > 115 characters)

File: corehq/apps/smsbillables/models.py [hunk 2]
```diff
@@ -94,10 +111,14 @@ def create_new(cls, backend_api_id, direction, amount,
 
 
     @classmethod
-    def get_by_criteria(cls, backend_api_id, direction, backend_instance=None, country_code=None):
-        criteria = SmsGatewayFeeCriteria.get_most_specific(backend_api_id, direction,
-                                                           backend_instance=backend_instance,
-                                                           country_code=country_code)
+    def get_by_criteria(cls, backend_api_id, direction, backend_instance=None, country_code=None, national_number=None):
```
Comment (dimagimon) [inline corehq/apps/smsbillables/models.py]: E501 line too long (120 > 115 characters)

**Comment by nickpell** [comment]
needs bootstrapping, tests, add column for prefix to sms gateway fee report

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 23 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 24 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 25 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 26 - E501 line too long (117 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 27 - E501 line too long (117 > 115 characters)

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 23 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 24 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 25 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 26 - E501 line too long (117 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 27 - E501 line too long (117 > 115 characters)

**Comment by dimagimon** [comment]
There are 9 errors:

* corehq/apps/smsbillables/management/commands/bootstrap_mach_gateway.py, line 31 - E231 missing whitespace after ','
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 23 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 24 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 25 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 26 - E501 line too long (117 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 27 - E501 line too long (117 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_tropo_gateway.py, line 18 - E303 too many blank lines (2)
* corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py, line 99 - E501 line too long (116 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py, line 100 - E501 line too long (120 > 115 characters)

**Comment by dimagimon** [comment]
There are 6 errors:

* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 23 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 24 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 25 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 26 - E501 line too long (117 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 27 - E501 line too long (117 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_tropo_gateway.py, line 18 - E303 too many blank lines (2)

**Comment by dimagimon** [comment]
There are 5 errors:

* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 23 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 24 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 25 - E501 line too long (118 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 26 - E501 line too long (117 > 115 characters)
* corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py, line 27 - E501 line too long (117 > 115 characters)

File: corehq/apps/smsbillables/generator.py [hunk 3]
```diff
@@ -146,3 +177,26 @@ def arbitrary_currency():
...
+def arbitrary_phone_numbers_and_prefixes(country_code_and_prefixes):
+    all_prefixes = [
+        prefix
+        for (country_code, prefix) in country_code_and_prefixes
+    ]
+    for country_code, prefix in country_code_and_prefixes:
+        remainder_len = 10 - len(prefix)
+
+        def _get_national_number():
+            return prefix + str(random.randint(10**(remainder_len - 1), 10**remainder_len - 1))
```
Comment (dimagimon) [inline corehq/apps/smsbillables/generator.py]: E226 missing whitespace around arithmetic operator

**Comment by dimagimon** [comment]
There are 3 errors:

* corehq/apps/smsbillables/generator.py, line 51 - E713 test for membership should be 'not in'
* corehq/apps/smsbillables/generator.py, line 189 - E713 test for membership should be 'not in'
* corehq/apps/smsbillables/generator.py, line 202 - E501 line too long (123 > 115 characters)

**Comment by dimagimon** [comment]
There are 3 errors:

* corehq/apps/smsbillables/generator.py, line 51 - E713 test for membership should be 'not in'
* corehq/apps/smsbillables/generator.py, line 189 - E713 test for membership should be 'not in'
* corehq/apps/smsbillables/generator.py, line 205 - E501 line too long (123 > 115 characters)

**Comment by nickpell** [comment]
closing in favor of https://github.com/dimagi/commcare-hq/pull/5618

**Pull request closed without merging**